thrust/thrust/system/cuda/detail/partition.h

```diff
@@ -38,1041 +38,349 @@
+  THRUST_INDEX_TYPE_DISPATCH2(
+    status,
+    dispatch32_t::dispatch,
+    dispatch64_t::dispatch,
+    num_items,
+    (policy, temp_storage, temp_storage_bytes, first, stencil, output, predicate, num_items_fixed, num_selected));
+  cuda_cub::throw_on_error(status, "partition failed on 2nd step");
+
+  return num_selected;
+}
 
-  template <typename Derived,
-            typename Iterator,
-            typename StencilIt,
-            typename Predicate>
-  THRUST_RUNTIME_FUNCTION
-  Iterator partition_inplace(execution_policy<Derived>& policy,
-                             Iterator                   first,
-                             Iterator                   last,
-                             StencilIt                  stencil,
-                             Predicate                  predicate)
-  {
-    typedef typename iterator_traits<Iterator>::difference_type size_type;
-    typedef typename iterator_traits<Iterator>::value_type      value_type;
-
-    size_type num_items = thrust::distance(first, last);
-
-    // Allocate temporary storage.
-    thrust::detail::temporary_array<value_type, Derived> tmp(policy, num_items);
-
-    cuda_cub::uninitialized_copy(policy, first, last, tmp.begin());
-
-    pair<Iterator, single_output_tag> result =
-        partition(policy,
-                  tmp.data().get(),
-                  tmp.data().get() + num_items,
-                  stencil,
-                  first,
-                  single_output_tag(),
-                  predicate);
-
-    size_type num_selected = result.first - first;
+template <typename Derived,
+          typename InputIt,
+          typename StencilIt,
+          typename SelectedOutIt,
+          typename RejectedOutIt,
+          typename Predicate>
+THRUST_RUNTIME_FUNCTION pair<SelectedOutIt, RejectedOutIt> stable_partition_copy(
+  execution_policy<Derived>& policy,
+  InputIt first,
+  InputIt last,
+  StencilIt stencil,
+  SelectedOutIt selected_result,
+  RejectedOutIt rejected_result,
+  Predicate predicate)
+{
+  using output_it_wrapper_t = cub::detail::partition_distinct_output_t<SelectedOutIt, RejectedOutIt>;
+  std::size_t num_items    = static_cast<std::size_t>(thrust::distance(first, last));
+  std::size_t num_selected = partition(
+    policy, first, last, stencil, output_it_wrapper_t{selected_result, rejected_result}, predicate);
+  return thrust::make_pair(selected_result + num_selected, rejected_result + num_items - num_selected);
+}
 
-    return first + num_selected;
-  }
-}    // namespace __partition
+template <typename Derived, typename InputIt, typename StencilIt, typename Predicate>
+THRUST_RUNTIME_FUNCTION InputIt inplace_partition(
+  execution_policy<Derived>& policy, InputIt first, InputIt last, StencilIt stencil, Predicate predicate)
+{
+  // Element type of the input iterator
+  using value_t         = typename iterator_traits<InputIt>::value_type;
+  std::size_t num_items = static_cast<std::size_t>(thrust::distance(first, last));
```

Comment:
question: I'm not sure about this cast. Temporary array checks for `num_items > 0`. If user accidentally provides last < first, we'll try to allocate size_t max. Do you think leaving num items signed until we reach dispatch is any better?

Reply:
Thanks, good catch on the negative ranges!

I've decided to check and return early for negative/empty ranges. I was concerned about other inconvenient side effects from negative `num_items`, like allocating the temp array with negative size for the `inplace` version, and returning the rejected iterator negative by the num items count (i.e., `rejected_result + num_items - num_selected`). So, instead we now have:
```
  if(thrust::distance(first, last) <= 0){
    return thrust::make_pair(selected_result, rejected_result);
  }
```
and 
```
  if(thrust::distance(first, last) <= 0){
    return first;
  }
```